src/librustc/traits/project.rs

```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```

Comment:
Hmm, I don't really expect you to be able to project `<u32 as Assoc>::Output` in an impl for `Assoc`. What I expect *could* work is if there is a DAG between the traits, sort of like this:

```rust
#![feature(specialization)]

use std::vec;

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <vec::IntoIter<u64> as Iterator>::Item {
    type Output = u64;
}

fn main() {
}
```

Interestingly, coherence rejects this example, presumably because of RFC 1214 concerns. I'm not sure that's right, but this example does work today:

```rust
#![feature(specialization)]

use std::vec;

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <u32 as Assoc2>::Output {
    type Output = u64;
}

trait Assoc2 {
    type Output;
}

impl Assoc2 for u32 {
    type Output = u64;
}

fn main() {
}
```

That said, I think that would also work in @michaelwoerister's branch, no?